File: CPP/MaxSumOfRectangleNoLargerThanK.cpp
#include <climits>
#include <iostream>
#include <set>
#include <vector>

class Solution {
public:
  int maxSumSubmatrix(std::vector<std::vector<int>> &matrix, int k) {
    int m = (int)matrix.size(), n = (int)matrix[0].size();
    long long best = LLONG_MIN;
    for (int left = 0; left < n; ++left) {
      std::vector<long long> rowsum(m, 0);
      for (int right = left; right < n; ++right) {
        for (int i = 0; i < m; ++i)
          rowsum[i] += matrix[i][right];
        std::set<long long> prefixes{0};
        long long cur = 0;
        for (long long v : rowsum) {
          cur += v;
          auto it = prefixes.lower_bound(cur - k);
          if (it != prefixes.end())
            best = std::max(best, cur - *it);
          prefixes.insert(cur);
        }
      }
    }
    return (int)best;
  }
};

int main() {
  Solution s;
  std::vector<std::vector<int>> a = {{1, 0, 1}, {0, -2, 3}};
  std::vector<std::vector<int>> b = {{2, 2, -1}};
  std::cout << s.maxSumSubmatrix(a, 2) << "\n"; // 2
  std::cout << s.maxSumSubmatrix(b, 3) << "\n"; // 3
  std::cout << s.maxSumSubmatrix(b, 0) << "\n"; // -1
  return 0;
}

File: Rust/src/bin/max_sum_of_rectangle_no_larger_than_k.rs
use std::collections::BTreeSet;

fn max_sum_submatrix(matrix: Vec<Vec<i32>>, k: i32) -> i32 {
    let m = matrix.len();
    let n = matrix[0].len();
    let k = k as i64;
    let mut best = i64::MIN;
    for left in 0..n {
        let mut rowsum = vec![0i64; m];
        for right in left..n {
            for i in 0..m {
                rowsum[i] += matrix[i][right] as i64;
            }
            // ordered prefix sums; smallest stored prefix >= cur - k
            let mut prefixes: BTreeSet<i64> = BTreeSet::new();
            prefixes.insert(0);
            let mut cur = 0i64;
            for &v in &rowsum {
                cur += v;
                if let Some(&p) = prefixes.range((cur - k)..).next() {
                    best = best.max(cur - p);
                }
                prefixes.insert(cur);
            }
        }
    }
    best as i32
}

fn main() {
    println!(
        "{}",
        max_sum_submatrix(vec![vec![1, 0, 1], vec![0, -2, 3]], 2)
    ); // 2
    println!("{}", max_sum_submatrix(vec![vec![2, 2, -1]], 3)); // 3
    println!("{}", max_sum_submatrix(vec![vec![2, 2, -1]], 0)); // -1
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn examples() {
        assert_eq!(max_sum_submatrix(vec![vec![1, 0, 1], vec![0, -2, 3]], 2), 2);
        assert_eq!(max_sum_submatrix(vec![vec![2, 2, -1]], 3), 3);
        assert_eq!(max_sum_submatrix(vec![vec![2, 2, -1]], 0), -1);
    }
}
